File: src/test/scala/com/github/kayjamlang/tests/expressions/operation/TwoFunctionOperationWithInFunctionContainerOperationTest.scala
package com.github.kayjamlang.tests.expressions.operation

import com.github.kayjamlang.core.{KayJamLexer, KayJamParser}
import com.github.kayjamlang.core.expressions.{CallOrCreateExpression, OperationExpression, ValueExpression}
import com.github.kayjamlang.core.expressions.data.Operation
import org.junit.Assert._
import org.junit.{BeforeClass, Test}


object TwoFunctionOperationWithInFunctionContainerOperationTest {
  private var parser: KayJamParser = null

  @BeforeClass def prepare(): Unit = {
    parser = new KayJamParser(new KayJamLexer("equals(2 - 1) + equals(1 + 1)"))
  }
}

class TwoFunctionOperationWithInFunctionContainerOperationTest {
  @Test
  @throws[Exception]
  def test(): Unit = {
    val expression = TwoFunctionOperationWithInFunctionContainerOperationTest.parser.readExpression
    assertNotNull(expression)
    assertSame(classOf[OperationExpression], expression getClass)
    val operationExpression = expression.asInstanceOf[OperationExpression]
    assertEquals(Operation.PLUS, operationExpression.operation)
    assertSame(classOf[CallOrCreateExpression], operationExpression.left getClass)
    assertSame(classOf[CallOrCreateExpression], operationExpression.right getClass)
    //Left
    val callOrCreateExpressionLeft = operationExpression.left.asInstanceOf[CallOrCreateExpression]
    assertEquals(1, callOrCreateExpressionLeft.arguments.size)
    assertSame(classOf[OperationExpression], callOrCreateExpressionLeft.arguments.head getClass)
    val operationExpressionLeft = callOrCreateExpressionLeft.arguments.head.asInstanceOf[OperationExpression]
    assertEquals(Operation.MINUS, operationExpressionLeft.operation)
    assertSame(classOf[ValueExpression], operationExpressionLeft.left getClass)
    assertSame(classOf[ValueExpression], operationExpressionLeft.right getClass)
    //Right
    val callOrCreateExpressionRight = operationExpression.right.asInstanceOf[CallOrCreateExpression]
    assertEquals(1, callOrCreateExpressionRight.arguments.size)
    assertSame(classOf[OperationExpression], callOrCreateExpressionRight.arguments.head getClass)
    val operationExpressionRight = callOrCreateExpressionRight.arguments.head.asInstanceOf[OperationExpression]
    assertEquals(Operation.PLUS, operationExpressionRight.operation)
    assertSame(classOf[ValueExpression], operationExpressionRight.left getClass)
    assertSame(classOf[ValueExpression], operationExpressionRight.right getClass)
  }
}

File: src/test/java/com/github/kayjamlang/tests/expressions/operation/TwoFunctionOperationWithInFunctionContainerOperationTest.java
package com.github.kayjamlang.tests.expressions.operation;

import com.github.kayjamlang.core.expressions.Expression;
import com.github.kayjamlang.core.KayJamLexer;
import com.github.kayjamlang.core.KayJamParser;
import com.github.kayjamlang.core.expressions.data.Operation;
import com.github.kayjamlang.core.expressions.CallOrCreateExpression;
import com.github.kayjamlang.core.expressions.ValueExpression;
import com.github.kayjamlang.core.expressions.OperationExpression;
import org.junit.BeforeClass;
import org.junit.Test;

import static org.junit.Assert.*;

public class TwoFunctionOperationWithInFunctionContainerOperationTest {

    private static KayJamParser parser;

    @BeforeClass
    public static void prepare(){
        parser = new KayJamParser(new KayJamLexer("equals(2-1)+equals(1+1)"));
    }

    @Test
    public void test() throws Exception {
        Expression expression = parser.readExpression();

        assertNotNull(expression);
        assertSame(OperationExpression.class, expression.getClass());

        OperationExpression operationExpression = (OperationExpression) expression;
        assertEquals(Operation.PLUS, operationExpression.operation);
        assertSame(CallOrCreateExpression.class, operationExpression.left.getClass());
        assertSame(CallOrCreateExpression.class, operationExpression.right.getClass());

        //Left
        CallOrCreateExpression callOrCreateExpressionLeft = (CallOrCreateExpression) operationExpression.left;
        assertEquals(1, callOrCreateExpressionLeft.arguments.size());
        assertSame(OperationExpression.class, callOrCreateExpressionLeft.arguments.get(0).getClass());

        OperationExpression operationExpressionLeft = (OperationExpression) callOrCreateExpressionLeft.arguments.get(0);
        assertEquals(Operation.MINUS, operationExpressionLeft.operation);
        assertSame(ValueExpression.class, operationExpressionLeft.left.getClass());
        assertSame(ValueExpression.class, operationExpressionLeft.right.getClass());

        //Right
        CallOrCreateExpression callOrCreateExpressionRight = (CallOrCreateExpression) operationExpression.right;
        assertEquals(1, callOrCreateExpressionRight.arguments.size());
        assertSame(OperationExpression.class, callOrCreateExpressionRight.arguments.get(0).getClass());

        OperationExpression operationExpressionRight = (OperationExpression) callOrCreateExpressionRight.arguments.get(0);
        assertEquals(Operation.PLUS, operationExpressionRight.operation);
        assertSame(ValueExpression.class, operationExpressionRight.left.getClass());
        assertSame(ValueExpression.class, operationExpressionRight.right.getClass());
    }
}
